### PR TITLE
vulkan: Fix Adreno crash for debug build

### LINKED ISSUE #5294: Adreno segfault with Vulkan when using unoptimized materials
Default settings and model, but must use a debug build to repro (unoptimized shaders are probably required).

The callstack contains `vkCreateGraphicsPipelines` which leads to `QGLCCompileToIRShader` in the qcom driver.

Pixel 4 with Android 12.

```
2022-03-08 14:33:23.350 13508-13528/com.google.android.filament.gltf I/AdrenoGLES-0: QUALCOMM build                   : 85da404, I46ff5fc46f
    Build Date                       : 11/30/20
    OpenGL ES Shader Compiler Version: EV031.31.04.01
    Local Branch                     : promo490_3_Google
    Remote Branch                    : 
    Remote Branch                    : 
    Reconstruct Branch               : 
2022-03-08 14:33:23.350 13508-13528/com.google.android.filament.gltf I/AdrenoGLES-0: Build Config                     : S P 10.0.4 AArch64
2022-03-08 14:33:23.350 13508-13528/com.google.android.filament.gltf I/AdrenoGLES-0: Driver Path                      : /vendor/lib64/egl/libGLESv2_adreno.so
2022-03-08 14:33:23.352 13508-13528/com.google.android.filament.gltf I/AdrenoGLES-0: PFP: 0x016ee190, ME: 0x00000000
```

```
2022-03-08 14:33:24.384 13508-13537/com.google.android.filament.gltf A/libc: Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x8 in tid 13537 (FEngine::loop), pid 13508 (d.filament.gltf)
2022-03-08 14:33:24.416 1113-1224/? E/QMI_FW: qmi_cci_get_ref: ref count increased 2
2022-03-08 14:33:24.416 1113-1224/? I/CHRE: @ 260338.395: [AR] inconsistent: 41.176471
2022-03-08 14:33:24.417 1113-1224/? E/QMI_FW: qmi_cci_get_ref: ref count increased 2
2022-03-08 14:33:24.417 1113-1224/? D/CHRE: @ 260338.396: [ActivityPlatform] type 6, confidence 41
2022-03-08 14:33:24.440 13555-13555/? I/crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstoneProto
2022-03-08 14:33:24.440 893-893/? I/tombstoned: received crash request for pid 13537
2022-03-08 14:33:24.441 13555-13555/? I/crash_dump64: performing dump of process 13508 (target tid = 13537)
2022-03-08 14:33:24.449 13555-13555/? E/DEBUG: failed to read /proc/uptime: Permission denied
2022-03-08 14:33:24.751 1896-3918/? W/InputManager-JNI: Input channel object '3522ad2 Toast (client)' was disposed without first being removed with the input manager!
2022-03-08 14:33:24.754 1896-3918/? W/NotificationService: Toast already killed. pkg=com.google.android.filament.gltf token=android.os.BinderProxy@eb3eea9
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: Build fingerprint: 'google/flame/flame:12/SQ1A.220205.002/8010174:user/release-keys'
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: Revision: 'MP1.0'
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: ABI: 'arm64'
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: Timestamp: 2022-03-08 14:33:24.449168353-0800
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: Process uptime: 0s
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: Cmdline: com.google.android.filament.gltf
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: pid: 13508, tid: 13537, name: FEngine::loop  >>> com.google.android.filament.gltf <<<
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: uid: 10465
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x8
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: Cause: null pointer dereference
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x0  b40000747cb8e0d0  x1  b40000734ca92190  x2  0000000000000001  x3  0000000000000045
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x4  b40000740cbd2520  x5  0000000000000005  x6  0000000000000021  x7  0000000000000001
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x8  0000000000000000  x9  874bac45304dc38d  x10 0000000000000000  x11 0000000000000021
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x12 0000000000000021  x13 b40000745ca55840  x14 000000727f882000  x15 0000000000000002
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x16 00000000000004c5  x17 00000072dc336118  x18 0000000000000007  x19 b4000073aca3ddb0
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x20 0000000000000000  x21 000000727f87e108  x22 3fffffffffffffff  x23 0000000000000004
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x24 0000000000000025  x25 b40000732cb128c0  x26 0000000000000004  x27 1fffffffffffffff
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     x28 0000000000000002  x29 b40000747cb8e0d0
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:     lr  00000072dce6efdc  sp  000000727f87df00  pc  00000072dcccae84  pst 0000000080000000
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG: backtrace:
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #00 pc 0000000000a7fe84  /vendor/lib64/libllvm-glnext.so (!!!0000!e996c91808ee26a66525968cb74e1e!85da404!+60) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #01 pc 0000000000c23fd8  /vendor/lib64/libllvm-glnext.so (!!!0000!5e4fdc544e8fb931f7dc5b11ff9a1d!85da404!+736) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #02 pc 0000000000c2483c  /vendor/lib64/libllvm-glnext.so (!!!0000!d7545dc46adf0a11a32e0f672f7e96!85da404!+972) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #03 pc 0000000000c3c0ec  /vendor/lib64/libllvm-glnext.so (!!!0000!d49599a40965508ac577c49e7c01e5!85da404!+348) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #04 pc 0000000000c3c5fc  /vendor/lib64/libllvm-glnext.so (!!!0000!f250bbeae4787d84d80d43e35a8b6d!85da404!+668) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #05 pc 0000000000c48290  /vendor/lib64/libllvm-glnext.so (!!!0000!d4b3104e6a37ce8c57eba970e8b2e1!85da404!+400) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #06 pc 0000000000a298a0  /vendor/lib64/libllvm-glnext.so (!!!0000!c03b43b7629d6f1414578265c7e520!85da404!+936) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #07 pc 0000000000b0fdc8  /vendor/lib64/libllvm-glnext.so (QGLCCompileToIRShader(void*, QGLC_SRCSHADER*, QGLC_COMPILETOIR_RESULT*)+88) (BuildId: 2e7eb8a8e15d025a6fb76421c6f455b6)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #08 pc 00000000000ae460  /vendor/lib64/hw/vulkan.adreno.so (!!!0000!408a0df45472c0e06e44d798996f89!85da404!+8248) (BuildId: 4eca9f7b16181c0ea2b281d9ab0f3c36)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #09 pc 00000000000943f0  /vendor/lib64/hw/vulkan.adreno.so (!!!0000!9edc5309b84ca97e813fe8bc838b43!85da404!+120) (BuildId: 4eca9f7b16181c0ea2b281d9ab0f3c36)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #10 pc 000000000006f2a4  /vendor/lib64/hw/vulkan.adreno.so (qglinternal::vkCreateGraphicsPipelines(VkDevice_T*, VkPipelineCache_T*, unsigned int, VkGraphicsPipelineCreateInfo const*, VkAllocationCallbacks const*, VkPipeline_T**)+2796) (BuildId: 4eca9f7b16181c0ea2b281d9ab0f3c36)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #11 pc 000000000050826c  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libVkLayer_khronos_validation.so (BuildId: 4dbe0f1b30e911d3f115344528a36a780368669d)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #12 pc 0000000000537a70  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libVkLayer_khronos_validation.so (BuildId: 4dbe0f1b30e911d3f115344528a36a780368669d)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #13 pc 0000000000604398  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #14 pc 0000000000603b7c  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #15 pc 000000000058fa6c  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #16 pc 00000000005b4160  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #17 pc 00000000005b4074  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #18 pc 00000000005b3fa8  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #19 pc 000000000058ef64  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #20 pc 0000000000588fd0  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #21 pc 000000000051c358  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #22 pc 000000000051c138  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.773 13555-13555/? A/DEBUG:       #23 pc 000000000051c0c8  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #24 pc 000000000051c068  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #25 pc 000000000051c038  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #26 pc 000000000051ae18  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #27 pc 00000000004892f4  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #28 pc 0000000000484534  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #29 pc 000000000051cc88  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #30 pc 000000000051a5d8  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #31 pc 00000000004408c4  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #32 pc 000000000043dd0c  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #33 pc 0000000000448484  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #34 pc 0000000000448368  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #35 pc 0000000000447b88  /data/app/~~0NgFBzu818ypKjfBxnA4Kg==/com.google.android.filament.gltf-WZSq7JNmYMsiUoZ8P9KmVw==/lib/arm64/libfilament-jni.so (BuildId: fc65b46e50d03b3b5eb094062e009441f7b5f8b0)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #36 pc 00000000000b1910  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+264) (BuildId: ba489d4985c0cf173209da67405662f9)
2022-03-08 14:33:24.774 13555-13555/? A/DEBUG:       #37 pc 00000000000513f0  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: ba489d4985c0cf173209da67405662f9)
```

### PR DESCRIPTION
**Pull request by poweifeng**
Remove `const` from `const highp mat4` just for mobile+vulkan
    
Fixes #5294

### CODE REVIEW

File: BUILDING.md [hunk 1]
```diff
@@ -270,6 +270,24 @@ This will generate Filament's Android binaries in `out/android-release`. This lo
 to build the Android Studio projects located in `filament/android`. After install, the library
 binaries should be found in `out/android-release/filament/lib/arm64-v8a`.
 
+#### Adreno workaround
+Adreno GPUs have issues using the optimized shaders compiled from Filament's materials. To
```
Comment (romainguy) [inline BUILDING.md:274]: Adrenos represent a large chunk of the population, we should really find a better workaround than asking devs to turn off optimizations entirely.

Comment (pixelflinger) [reply BUILDING.md:274]: We tried!

File: shaders/src/common_defines.glsl [hunk 2]
```diff
@@ -22,3 +22,13 @@
...
+
+// To workaround an adreno crash (#5294), we need ensure that a method with
+// parameter 'const mat4' does not call another method also with a 'const mat4'
+// parameter (i.e. mulMat4x4Float3). So we remove the const modifier for
+// materials compiled for vulkan+mobile.
+#if defined(TARGET_VULKAN_ENVIRONMENT) && defined(TARGET_MOBILE)
+   #define highp_mat4 highp mat4
+#else
+   #define highp_mat4 const highp mat4
+#endif
```
Comment (pixelflinger) [inline shaders/src/common_defines.glsl:34]: what about we do this for everything everywhere and all types?

also, do you think we could do this at the SPIRV level? i.e. could we patch the SPIRV only on Adreno?

I think we should have a small spirv util lib we can use to manipulate spirv, I think it's going to become needed anyways.

File: BUILDING.md (hunk 1 above)
Comment (romainguy) [reply BUILDING.md:274]: Then we should at least say it's specific to Vulkan?

Comment (poweifeng) [reply BUILDING.md]: no longer required as it's separately addressed in #6757

File: shaders/src/common_defines.glsl (hunk 2 above)
Comment (poweifeng) [reply shaders/src/common_defines.glsl:34]: Do you mean take out all of the const modifier to the parameters? 

So interestingly, taking out the const in this case does seem to result in one extra store and one load in the resulting spirv. Though that might get optimized away through the optimizer. But I still hesitate to make this a larger change than necessary.

**Review by pixelflinger — APPROVED** [review]

**Pull request merged**